planner/cascades/optimize.go

```diff
@@ -23,6 +23,30 @@ import (
 	"github.com/pingcap/tidb/sessionctx"
 )
 
+// Optimizer is the struct for cascades optimizer.
+type Optimizer struct {
+	transformationRuleMap map[memo.Operand][]Transformation
+	implementationRuleMap map[memo.Operand][]ImplementationRule
+}
+
+// NewDefaultOptimizer returns a cascades optimizer with default transformation
+// rules and implementation rules.
+func NewDefaultOptimizer() *Optimizer {
+	return &Optimizer{
+		transformationRuleMap: defaultTransformationMap,
+		implementationRuleMap: defaultImplementationMap,
+	}
+}
+
+// NewOptimizerWithTrans returns a cascades optimizer which uses user defined
+// transformation rule map. This will be frequently used in unit tests.
+func NewOptimizerWithTrans(trans map[memo.Operand][]Transformation) *Optimizer {
```

Comment:
How about adding the following two functions for the `Optimizer` struct:
```go
func (opt *Optimizer) ResetTransformationRules(map[memo.Operand][]Transformation) { ... }
func (opt *Optimizer) ResetImplementationRules(map[memo.Operand][]ImplementationRule) { ... }
```

Reply:
Looks good.
It's convenient for unit tests, then we only need one `Optimizer` in a test suite, and we can reset the rules in each unit tests.